content/lectures/lec02.md

```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
+
+We can also build the image locally and send the image after it has already been built. This way the receiver can run the container without having to worry about any of the build process.
+
+Ideally, the build process itself should also be reproducible. 
+
+## Docker
+
+Docker is the industry-standard implementation for containers. While there are other container implementations, Docker has become something like the "Kindle" or "Kleenex" of container implementations, but there [are](https://coreos.com/rkt/) [other](https://katacontainers.io) [implementations](https://firecracker-microvm.github.io) which have emerged recently and look to be giving Docker a run for its money. 
+
+The Open Container Image (OCI) format is a general specification for how to build container images originally based on the Docker image format. But other tools built against the OCI specification build compatible containers or run compatible containers, without running any code built by the Docker team. Implementations like [podman](https://containerd.io) and [containerd](https://containerd.io) are compatible with the OCI specification while improving on some ambiguities in the original Docker implementation. Kubernetes has actually [moved away from the Docker implementation](https://kubernetes.io/blog/2020/12/02/dont-panic-kubernetes-and-docker/) recently. We'll use "OCI container", "container", and "Docker container" interchangeably throughout these notes.
+
+## Virtual Machines vs. Containers
+
+For each virtual machine, the hypervisor fully emulate every piece of hardware that exists on the machine. Every virtual machine does this emulation individually, so if we have many virtual machines they don't share the majority of their resources. VMs each run their own kernel to talk to the emulated hardware.
+
+![Virtual Machine Diagram](/img/lec02/vm_vis.png)
+
+Instead, containers share as many resources between multiple containers and with the host OS as possible (while still maintaining privacy and separation). This gives us a huge performance increase because emulation is always slower than actual computation. This importantly includes the host OS's kernel, which talks to the real hardware on the computer. This makes Docker containers much lighter weight than equivalent VMs.
+
+![Container Diagram](/img/lec02/container_vis.png)
+
+## Docker Images
+
+Docker images ideally contain the minimal OS needed to run your application. We only want the programs that are needed to run our application and nothing more.
+
+Docker images are built in layers. When we create a Docker image, it is always extending a base image. This is achieved with a "copy-on-write" filesystem for each image. This means that the filesystem within the image is completely immutable during the build process. Instead if modifying a file in-place, Docker will create an additional image layer with your changes on top of the base image.
```

Comment:
```suggestion
Docker images are built in layers. When we create a Docker image, it is always extending a base image. This is achieved with a "copy-on-write" filesystem for each image. This means that the filesystem within the image is completely immutable during the build process. Instead of modifying a file in-place, Docker will create an additional image layer with your changes on top of the base image.
```